communication/include/communication/StateControl.hpp

```diff
@@ -14,31 +14,89 @@
 #include <functional>
 #include <utility>
 #include "callback_registry.hpp"
+#include <unordered_map>
+#include <variant>
+#include <memory>
+#include "Logger.hpp"
+
 namespace bft::communication {
+struct EnumHash {
+  template <typename T>
+  size_t operator()(T t) const {
+    return static_cast<size_t>(t);
+  }
+};
+
 class StateControl {
  public:
+  using CallbackRegistry = concord::util::CallbackRegistry<uint32_t>;
+  enum class EventType { TLS_COMM, THIN_REPLICA_SERVER };
+
   static StateControl& instance() {
     static StateControl instance_;
     return instance_;
   }
+
   void lockComm() { lock_comm_.lock(); }
   bool tryLockComm() { return lock_comm_.try_lock(); }
   void unlockComm() { lock_comm_.unlock(); }
 
-  void setCommRestartCallBack(std::function<void(uint32_t)> cb) {
-    if (cb != nullptr) comm_restart_cb_registry_.add(cb);
+  // Added/retained these methods to act as facade
+  void setCommRestartCallBack(std::function<void(uint32_t)> cb, const EventType& et = EventType::TLS_COMM) {
+    registerCallback(et, cb);
+  }
+
+  void restartComm(uint32_t id, const EventType& et = EventType::TLS_COMM) { invokeCallback(et, id); }
+
+  void setTlsRestartCallBack(std::function<void(uint32_t)> cb, const EventType& et = EventType::THIN_REPLICA_SERVER) {
+    registerCallback(et, cb);
   }
 
-  void restartComm(uint32_t id) { comm_restart_cb_registry_.invokeAll(id); }
+  void restartThinReplicaServer(uint32_t id, const EventType& et = EventType::THIN_REPLICA_SERVER) {
+    invokeCallback(et, id);
+  }
+
+  const std::string getEventTypeAsString(const EventType& et) {
+    switch (et) {
+      case EventType::TLS_COMM:
+        return "TLS_COMM";
+      case EventType::THIN_REPLICA_SERVER:
+        return "THIN_REPLICA_SERVER";
+      default:
+        return "Invalid Type";
+    }
+  }
   void setGetPeerPubKeyMethod(std::function<std::string(uint32_t)> m) { get_peer_pub_key_ = std::move(m); }
+
   std::string getPeerPubKey(uint32_t id) {
     if (get_peer_pub_key_) return get_peer_pub_key_(id);
     return std::string();
   }
 
  private:
+  StateControl() : logger_(logging::getLogger("concord-bft.comm.StetControl")) {}
+
+  void registerCallback(const EventType& et, std::function<void(uint32_t)> cb) {
+    if (cb != nullptr) {
```

Comment:
what is the use case for a null callback passed?
If there is a case can you please shortly explain?
If not use-case, you should assert here instead.

Reply:
added assert